cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp

```diff
@@ -0,0 +1,91 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#include "IdleTimeoutTransceiverDecorator.h"
+#include "Ice/Buffer.h"
+
+using namespace std;
+using namespace Ice;
+using namespace IceInternal;
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::initialize(Buffer& readBuffer, Buffer& writeBuffer)
+{
+    SocketOperation op = _decoratee->initialize(readBuffer, writeBuffer);
+
+    if (op == SocketOperationNone) // connected
+    {
+        rescheduleKeepAlive();
+    }
+    return op;
+}
+
+IdleTimeoutTransceiverDecorator::~IdleTimeoutTransceiverDecorator() { _timer->cancel(_keepAliveAction); }
+
+void
+IdleTimeoutTransceiverDecorator::close()
+{
+    _timer->cancel(_keepAliveAction);
+    _decoratee->close();
+}
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::write(Buffer& buf)
+{
+    Buffer::Container::iterator start = buf.i;
+    SocketOperation op = _decoratee->write(buf);
+    if (buf.i != start)
+    {
+        // We've sent at least one byte, reschedule the keep alive.
```

Comment:
I used the tracing code in ConnectionI.cpp to figure out when we write something. I hope this tracing code is complete.